### ISSUE TITLE
Performance degradation from 11.1.5 to 14.10.2

### ISSUE THREAD

**Issue by rogerta** [tapjs/tapjs#633]
### TL;DR
I recently upgraded from tap 11.1.5 to 14.10.2 and see ~200% slow down in test performance.  I read through issue #457 so I realize there are known problems, but I was wondering if you have any advice for speeding things up in the mean time.

23 test suites, 2117 assertions, tap v11.1.5: average run time of ~13s.
23 test suites, 2117 assertions, tap v14.10.2: average run time of ~31s.

### Long version
node: v12.13.1
npm: v6.12.1
tap: v11.1.5 and v14.10.2
OS: Linux penguin 4.19.69-06666-g6c4f8cbba24e #1 SMP PREEMPT Fri Sep 6 22:15:24 PDT 2019 x86_64 GNU/Linux
Hardware: Pixelbook
Google Chrome | 78.0.3904.106 (Official Build) (64-bit)
Revision | 8543c99173fbc36cf5ce3019f07b7ab3539ca10f-refs/branch-heads/3904@{#879}
Platform | 12499.66.0 (Official Build) stable-channel eve
Firmware Version | Google_Eve.9584.174.0
Customization ID | GOOGLE-EVE

My old package.json:
```javascript
{
  ...
  "dependencies": {
    "@webcomponents/webcomponentsjs": "^2.4.0",
    "lit-element": "^2.2.1"
  },
  "devDependencies": {
    "@types/gapi": "^0.0.35",
    "@types/gapi.client.drive": "^3.0.1",
    "@types/node": "^12.7.2",
    "@vaadin/vaadin-icons": "4.2.0-alpha1",
    "byline": "^5.0.0",
    "polymer-cli": "^1.9.11",
    "tap": "^11.0.1",
    "typescript": "^3.6.3",
    "workbox-cli": "^4.3.1"
  },
  "resolutions": {
    "@types/form-data": "2.2.1"
  }
}
```

My new package.json  is the same as above but replaces the tap line with:
```javascript
    "tap": "^14.10.2",
```

With the old package.json, I run tests as follows:

```sh
$ npx tap -J --cov out/tests/**/*_tests.js
...
total ........................................... 2117/2117

  2117 passing (11s)

  ok
------------------------|----------|----------|----------|----------|-------------------|
File                    |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------|----------|----------|----------|----------|-------------------|
All files               |      100 |      100 |      100 |      100 |                   |
...
```

With the new package.json, I run tests as follows:

```sh
$ npx tap --changed out/tests/**/*_tests.js

  🌈 SUMMARY RESULTS 🌈  

Suites:   23 passed, 23 of 23 completed
Asserts:  2117 passed, of 2117
Time:     31s
------------------------|----------|----------|----------|----------|-------------------|
File                    |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------|----------|----------|----------|----------|-------------------|
All files               |      100 |      100 |      100 |      100 |                   |
...
```

I ran the tests 6 times in a row with tap v11.1.5 and got times of 28s, 11s, 10s, 10s, 11s, 10s.
I then ran the tests 5 times in a row with tap v14.10.2 and got times of 29s, 31s, 31s, 31s, 31s.
In both cases no TAP_xxx environment variables were set and no `.taprc` file was used.

With v14 I didn't use `-J --cov` because those are on by default.  I find it particularly strange that while I used `--changed` the run times remained essentially the same, even though no code changes were done between test runs.  Am I misusing the command line arg?

While the source code is written in typescript, it gets compiled to commonjs javascript in the out/tests directory.  Tap runs the commonjs javascript and is not doing any typescript compiling.

## Ways to reduce test run times in v14
I was able to bring down the run times with the `-Rtap` command line argument: 23s, 27s, 26s, 27s, 24s.  So it seems the fancy UI  cost ~15% of running time.

Also, running `npx tap --watch out/tests/**/*_tests.js` still takes ~31s to run tests the first time but runs much more quickly afterwards.  A one file change seems to be detected in a little over 3s and runs in a little over 2s for a total time of ~6s.

Any other recommendations?

**Comment by coreyfarrell** [comment]
Could you try using `--no-esm`?  According to the changelog running with the `esm` module was added to tap 12.5.0 and it is enabled by default.

**Comment by rogerta** [comment]
Thanks for the suggestion @coreyfarrell, that made quite a difference.  So when I run tests with the command `npx tap --no-esm --changed out/test/**/*_tests.js`, v14 is only ~90% slower than v11: run times of 21s, 20s, 20s, 20s, 20s.

If I combine your suggestion with `-Rtap` v14 is only ~50% slower than v11: 14s, 18s, 18s, 17s, 16s.

**Comment by coreyfarrell** [comment]
Another potential source could be nyc.  tap 11.x used nyc 11.x which was based on babel 6.  tap 14 uses nyc 14 which is based on babel 7.  Would be interesting to see the difference with coverage disabled on both versions of tap.

**Comment by isaacs** [comment]
@rogerta re the `--changed` flag, it isn't relevant if you are specifying a file list.  `--changed` will tell tap to build up the file list based on what files changed since the last run, but if you're specifying a file list, it will always run the files specified.

**Comment by rogerta** [comment]
@coreyfarrell I think your suspicions are correct.  When I run both v11 and v14 with the `--no-cov` command line arg the running times are essentially the same:

v11: 13s, 13s, 13s, 13s, 13s
v14: 13s, 13s, 17s, 14s, 14s

Coverage is important to me so I won't be running tests with `--no-cov` but just wanted to give you some data.  Note that when I run `npx tap --watch out/tests/**/*_tests.js` this most often results in shorter run times with v14 than my previous workflow with v11 (which was run all tests all the time, I didn't realize v11 also supported `--watch`), so I will stick with v14 moving forward.  Looking forward to even faster coverage with c8 :-)

@isaacs thanks for the info about `--changed`.

**Comment by coreyfarrell** [comment]
@rogerta Thanks for running all these tests.

You can actually use c8 now if you want.  For this you have to install c8 and run `npx c8 tap --no-cov`.  I've been told by the author of c8 that performance of c8 vs nyc can go either way.  He believes this may have to do with v8 potentially disabling some optimizations when c8 is running.

Also worth noting nyc 15 might bring some performance enhancement for tests with many test suites (lower per process overhead).  To try nyc 15 with tap:
1. `npm i -D nyc@next`
2. Alter your script to run `nyc tap --no-cov` in addition to your other tap arguments.

I'd be interested in knowing if you see any performance benefit to `nyc tap --no-cov` (nyc 15).  I would not expect much difference for the first run of your tests but I would expect repeat runs to be faster with nyc 15.

**Comment by rogerta** [comment]
@coreyfarrell here are some more results using straight tap as well as tap with nyc and tap with c8.

### tap@14.10.2 (implicit nyc@14.1.1)
Command: `npx tap out/tests/**/*_tests.js`
Run times: 48s 32s 33s 33s 33s 33s

### tap@14.10.2 + nyc@15.0.0-beta.3
Command: `npx nyc tap --no-cov out/tests/**/*_tests.js`
Run times: 30s 25s 22s 21s 21s 20s

### tap@14.10.2 + c8@6.0.1
Command: `npx c8 tap --no-cov out/tests/**/*_tests.js`
Run times: 17s 20s 19s 19s 20s 19s

So it seems using tap with either c8 or a newer nyc is faster than using straight tap alone.  However, there is some strange behaviour with c8:

1. There was ~3s pause between tap showing "SUMMARY RESULTS" and the command prompt becoming responsive again.  If you add this to the c8 run times then it's really no better than nyc@15.  All run times reported here and all my previous posts are the times reported by tap.
1. c8 determines I don't have 100% function coverage whereas tap alone and tap+nyc both say I do.  In two cases c8 does not indicate uncovered lines at all so I can't tell which lines it thinks are uncovered.  In three other cases c8 does indicate uncovered lines, but those lines don't contain instructions;  they are in the middle of a literal array declaration.

I'm not sure if this c8 test is valid.

**Comment by coreyfarrell** [comment]
So tap@14.10.2 + nyc@15.0.0-beta.3 with `--no-esm` passed to tap likely gives similar performance to tap@11.x?  Maybe a little slower due to the nicer reporting.  The plan is that tap@15 will use nyc 15 and `--no-esm` will be the default, so it will have better performance.

As for measuring the whole process, I suggest using the `time` utility (sorry I didn't think of this earlier).  As an example you would run `time npx nyc tap --no-cov --no-esm out/tests/**/*_tests.js`.  Even in the case of nyc it performs some processing outside of the time reported by tap to merge and report coverage results.  Of course to compare multiple scenarios (c8 vs nyc for example) you'd have to compare `time` to `time`, you cannot compare the result of `time` to what tap reports for test runtime.

Another thing I'll point out in the timing results, the first run with nyc 15 is slow compared to all other runs.  This is not an anomaly, the first run transforms your source code to add coverage counters, caches the transformed code to `node_modules/.cache/nyc`.  The additional runs of nyc then find the cached file and skips the transform.  nyc 14 unconditionally loads babel even though it may go unused, nyc 15 doesn't load babel until a cache miss occurs.

I can't answer about the c8 issue, it sounds like a source-map is not being picked up or was processed improperly.  If you are interested in c8 I'd suggest trying `c8@next` (7.0.0-candidate.0).  If this still does not give results you expect I'm sure a bug report would be appreciated at https://github.com/bcoe/c8, especially if you could provide a minimal reproduction.
